frontend/src/metabase/query_builder/components/filters/modals/BulkFilterModal/utils.ts

```diff
@@ -0,0 +1,63 @@
+import StructuredQuery from "metabase-lib/lib/queries/StructuredQuery";
+import Filter from "metabase-lib/lib/queries/structured/Filter";
+
+// fix between filters with missing or misordered arguments
+export function fixBetweens(query: StructuredQuery): StructuredQuery {
+  const betweenFilters = query
+    .filters()
+    .filter(filter => filter.operatorName() === "between");
+
+  for (const filter of betweenFilters) {
+    const validArguments = countValidArumgents(filter);
+    if (validArguments === 1) {
+      return fixBetweens(handleEmptyBetween(filter));
```

Comment:
If you return here, you will not call `fixBetweens` on later filters when there are multiple filters from the notebook. 
```suggestion
      query = fixBetweens(handleEmptyBetween(filter));
```

Reply:
I don't think that will actually work. Bear with me for a minute, the way queries and filters are interrelated is a little mind-bending.

If we reassign the query as you suggest, we also have to re-extract the filters, because they all have the query inside them. If we don't re-extract the filters from the new query, calling `filter.replace()` will replace the filter in the old query.

This is why I think it made the most sense to make this function recursive. On each recursion, only one filter can get replaced, and then we start over with a new query and its own extracted set of filters. We reach the base case when we reach a query where there are no longer any filters with fixable problems.

Reply:
Upon further investigation, your solution also works, but I think it's less efficient.

If you run the last unit test for this file, it tests this exact case, where there are multiple filters that need correction. If we console log every iteration of the loop, the one with the early return loops 4 times, where reassigning the variable loops 7 times.  Those extra iterations are unnecessary because we can still only replace one filter at each level of recursion, so returning early keeps us from doing unnecessary work.

Reply:
@iethree you were right. I was just imagining things. I retested and everything works as expected.

One problem with this code though that it's not apparent it's a recursive function, I mean I kind of noticed you call `fixBetweens` recursively, but I got distracted by how you're looping through `betweenFilters` and only return once. I mean I would expect to see some kind of reduced work in each recursive loop. With this code, we have to loop through `betweenFilters` from the start every time, when we already know how much we've progressed.

I tried changing the last unit test for `fixBetweens` so that it contains 3 `between` filter that needs to be swapped. And log at the start of the loop and it logs 9 times.

```ts
      const testQuery = makeQuery({
        filter: [
          "and",
          ["between", ["field", ORDERS.fields[1].id, null], 80, 50],
          ["between", ["field", ORDERS.fields[1].id, null], 80, 50],
          ["between", ["field", ORDERS.fields[1].id, null], 80, 50],
          ["=", ["field", ORDERS.fields[1].id, null], 8, 9, 7],
        ],
      });
```

1st iteration: loop 1 time before finding the next between filter that needs to be fixed
2nd iteration: loop 2 times before finding the next between filter that needs to be fixed
3rd iteration: loop 3 times before finding the next between filter that needs to be fixed
4th iteration: loop 3 times again but don't find anything to fix as all filters have been fixed

I'd expect a recursive function to know the state at which it needs to loop e.g.

```ts
export function fixBetweens(
  query: StructuredQuery,
  loopIndex: number = 0,
): StructuredQuery {
  const betweenFilters = query
    .filters()
    .filter(filter => filter.operatorName() === "between");

  const filter = betweenFilters[loopIndex];
  if (filter) {
    const validArguments = countValidArumgents(filter);
    if (validArguments === 1) {
      return fixBetweens(handleEmptyBetween(filter), loopIndex + 1);
    } else if (validArguments === 2 && hasBackwardsArguments(filter)) {
      return fixBetweens(swapFilterArguments(filter), loopIndex + 1);
    }
  }

  return query;
}
```

What do you think?

Reply:
I don't think that works because the `loopIndex` is not stable because we potentially have a different number of `between` filters on each level of recursion because we're changing some of the `between` filters to `>=` or `<=`.  Your example works because we're not changing any operators there, but it won't work for the empty argument case.

I do see what you mean about the combination of looping + recursion being confusing. One alternative I can think of is using a `.find()` to make it clear that we're just looking for a single item. The problem with this is that we're duplicating some code and work: namely counting arguments an extra time.

```js
export function fixBetweens(query: StructuredQuery): StructuredQuery {
  const betweenFilters = query
    .filters()
    .filter(filter => filter.operatorName() === "between");

  const firstInvalidFilter = betweenFilters.find((filter) => {
    const validArgumentsCount = countValidArumgents(filter);
    return (validArgumentsCount === 1) || (validArgumentsCount === 2 && hasBackwardsArguments(filter));
  });

  if (firstInvalidFilter) {
    const validArgumentsCount = countValidArumgents(firstInvalidFilter);
    return fixBetweens(
      validArgumentsCount === 1
        ? handleEmptyBetween(firstInvalidFilter)
        : swapFilterArguments(firstInvalidFilter)
    );
  }

  return query;
}
```

Functionally, we're doing the same thing each time: finding the first invalid filter that we can fix, and fixing it. I'm also not sure if we can always rely on having filters in a stable order, and even if we can now, I'd rather not build that assumption in, because it might not be obvious down the road that there are pieces of the codebase that rely on a stable filter order.

Reply:
That's fair. But from `fixBetweens`'s point of view the filter order should be stable right? As each operation on the filter/query is pure, `fixBetweens`'s job can guarantee not to, for example, swap any filter order that would mess up the loop. And isn't not possible for other part of the code to tamper with `query` argument as (as far as I understand the code) all operations on `Filter` and `Query` are pure (no mutation).

Sorry, that I might seem stubborn to get the loop working 😅 , but I just want to validate if my understanding of `Filter` and `Query` is correct.

Do you think the following code will work?

```
// fix between filters with missing or misordered arguments
export function fixBetweens(
  query: StructuredQuery,
  loopIndex = 0,
): StructuredQuery {
  const filters = query.filters();

  // find the first invalid filter (if any), and fix it recursively
  const filter = filters[loopIndex];
  if (filter && filter.operatorName() === "between") {
    const validArgumentsCount = countValidArumgents(filter);
    if (validArgumentsCount === 1) {
      return fixBetweens(handleEmptyBetween(filter), loopIndex + 1);
    } else if (validArgumentsCount === 2 && hasBackwardsArguments(filter)) {
      return fixBetweens(swapFilterArguments(filter), loopIndex + 1);
    }
  }

  return query;
}
```

Based on my assumption that the order is guaranteed, I don' filter the list before so filters should be in the same order (as `fixBetweens` doesn't change the filters order)